### ISSUE TITLE
Processes Displayed as Trees Auto Collapse on Change 

### ISSUE THREAD

**Issue by rt-kill** [hakandundar34coding/system-monitoring-center#154]
I prefer viewing the processes as trees instead of a giant mess of individual items. I see that there is an option to view processes as trees, however they continue to auto collapse. I'll will click to open a fold, and then it will immediately close afterwards. Is there a fix for this?

**Comment by amiga-500** [comment]
Interesting.  What OS are u using?  I'm using Ubuntu and it doesn't have this problem (when i switch to trees they are open and stay open even when i navigate to a diff tab and back -- selectively collapsing is also remembered...not on re-launch ofc).

For me, although i use the processes tab a lot, i still prefer non-trees.  I think because i depend on sorting by CPU% which is made more difficult with trees in situations.   But i can def see the appeal by others for that feature.

**Comment by hakandundar34coding** [comment]
There is no similar tree bug on my system. Can you give more details about the problem? Can you share a screenshot of the System tab of the application for your system details.

New processes are shown as collapsed if they contain child processes. Are your processes ended and started again?

**Comment by rt-kill** [comment]
## Screen Recording
Here is a screen recording I made. This was the original behavior I noticed after my initial install, but I was not the worst thing to deal with.

Yesterday it was doing the opposite where trees were default collapsed, and shortly after opening one it would re-close. This made the tree tabs unusable. Since then I have run a system update and don't know how this might play into things.

In this video you can see I switch Workspaces and open a terminal. This is only for trying to reproduce the problem, even though it has happened without me leaving the window.

https://user-images.githubusercontent.com/69824969/232329229-978b7995-5901-41f1-b81e-38a380417af9.mp4

## System
Here is a screen shot of the System tab

![image](https://user-images.githubusercontent.com/69824969/232329767-ed6ed41a-f9a2-4bac-bbef-ad539de371a1.png)

Additionally, I am using i3wm.

## Continuation
> New processes are shown as collapsed if they contain child processes. Are your processes ended and started again?

I wouldn't think so, as the entire kthreadd and systemd  trees are closing.

> For me, although i use the processes tab a lot, i still prefer non-trees. I think because i depend on sorting by CPU% which is made more difficult with trees in situations. But i can def see the appeal by others for that feature.

In general I agree that sorting by CPU/Memory is the way to go, but I like to know where the process is starting from and what child processes exist. The specific example that landed me here was that I was building a lot of orphaned processes that were consuming large amounts of cpu, but didn't know where they came from. With the tree tabs, I was able to find the problem.

**Comment by hakandundar34coding** [comment]
Do you use keyboard buttons to change windows?
Expand all and Collapse all buttons in the customization pop up menu may be activated during window changes.

How do you close the processes customization pop up menu?

**Comment by rt-kill** [comment]
I do use keyboard shortcuts to change windows, but have not found any patterns linking them to the tree collapse's. The trees have collapsed without any change of windows. 

To close the processes customization pop up, i use i3wm's kill option to kill the currently focused window (via keybindings), but again tree collapses have happened without any killing of windows.

**Comment by hakandundar34coding** [comment]
I found the problem. Running a new process after collapsing the tree expands it again. System Monitoring Center runs new processes in the background for getting process information. Tree is expanded again even if the user does not run any applications.

The bug is fixed. Can you follow these steps for testing?

- Download the [latest](https://github.com/hakandundar34coding/system-monitoring-center/archive/refs/heads/master.zip) source code
- Extract it and open terminal in `src` folder
- Run this command: `python3 ./run_from_source.py`
- Check the tree behavior

**Comment by rt-kill** [comment]
Awesome!

Here's a video of me running `python3 ./run_from_source.py`:

https://user-images.githubusercontent.com/69824969/232552687-948fc1a5-4a50-4a8c-ab3b-4eee2b3e9274.mp4

On separate attempts, I was receiving the other following errors:

```
Traceback (most recent call last):
  File "/home/rolandtk/Downloads/sysmon/system-monitoring-center-master/src/Processes.py", line 1267, in processes_loop_func
    self.piter_list.append(self.treestore.append(self.piter_list[pid_list.index(ppid_list[pid_list.index(process)])], processes_data_rows[pid_list.index(process)]))
IndexError: list index out of range
```

```
Traceback (most recent call last):
  File "/home/rolandtk/Downloads/sysmon/system-monitoring-center-master/src/Processes.py", line 714, in on_treeview_pressed
    self.selected_process_pid = self.pid_list[self.processes_data_rows.index(model[treeiter][:])]
AttributeError: 'Processes' object has no attribute 'pid_list'. Did you mean: 'piter_list'?
```

I'm not familiar with the code, and can't really understand these errors. I did try for a little bit to trace them, but didn't get very far.

**Comment by hakandundar34coding** [comment]
I will make more detailed code changes and write here the result.

**Comment by rt-kill** [comment]
Sounds good.

**Comment by hakandundar34coding** [comment]
I have updated the source code. Can you follow these steps for testing it?

- Download the [latest](https://github.com/hakandundar34coding/system-monitoring-center/archive/refs/heads/master.zip) source code
- Extract it and open terminal in `src` folder
- Run this command: `python3 ./run_from_source.py`
- Check the tree behavior and command output for errors.

Note: There will be minor improvements.

**Comment by rt-kill** [comment]
Cool yea, I notice some minor improvements -- mainly that that the root of trees are no longer auto opening/closing. The sub trees, however, are still auto opening. I don't know if this is what you meant by minor improvements.

**Comment by hakandundar34coding** [comment]
The code is updated again. After the change, sub-trees are not expanded automatically.

**Comment by hakandundar34coding** [comment]
The code and the application are updated. Can you test the [latest](https://github.com/hakandundar34coding/system-monitoring-center/issues/154#issuecomment-1514976484) source code or the latest version (v2.11.0) if it is added to the repository?

**Comment by rt-kill** [comment]
It looks like sub-trees are no longer collapsing. The process, however, did crash after about 30 seconds (both times I launched it) with the following errors:

First run:
![image](https://user-images.githubusercontent.com/69824969/233694424-63490a57-2971-42c2-a514-2ff4edda4544.png)

Second run:
![image](https://user-images.githubusercontent.com/69824969/233694475-03c0833f-0bc8-4bd3-8a26-d9ec8f7ae7d6.png)

**Comment by hakandundar34coding** [comment]
Does it crash even if the search box is empty or processes are not listed as tree?

**Comment by rt-kill** [comment]
It crashes even with the search box empty (with the processes listed as trees).

It does not seem to crash when the processes are not listed as trees (even when using the search box).

**Comment by hakandundar34coding** [comment]
I tested the application several times. It crashed two times. Additionally, there are several warning messages in the output.
The behavior may be different on other systems.

Probably, treeview widget will be replaced by new columnview. There may be a lot of code changes. This may take more than a few weeks. The old widget is deprecated. I do not want to spend more time for fixing the bug in the code that uses this old widget.

**Comment by rt-kill** [comment]
👍 Understood

**Comment by hakandundar34coding** [comment]
The bug is fixed and the application is updated (v2.14.0). It may be downloadable from:

- Flathub in about 4 hours
- Other repositories in about a few days

Treeview widget is not replaced by new columnview. Because CPU usage was higher during tests.

I will close this issue. But you can write additional comments if there are problems about treeview expanding/collapsing/searching behavior.

**Comment by rt-kill** [comment]
Looking forward to it 👍
